cabal-install-solver/src/Distribution/Solver/Modular/Message.hs

```diff
@@ -14,124 +14,166 @@ import qualified Data.Set as S
+      ConflictSet,
+      showConflictSet,
+      QGoalReason,
+      GoalReason(DependencyGoal, UserGoal),
+      Goal(Goal),
+      DependencyReason(DependencyReason),
+      ExposedComponent(..),
+      PkgComponent(PkgComponent),
+      CI(Constrained, Fixed),
+      showDependencyReason )
+import Distribution.Solver.Modular.Flag
+    ( showQFNBool, showQSNBool, showQFN, showQSN )
 import Distribution.Solver.Modular.MessageUtils
-         (showUnsupportedExtension, showUnsupportedLanguage)
+    ( showUnsupportedExtension, showUnsupportedLanguage )
 import Distribution.Solver.Modular.Package
+    ( PI(PI), showI, showPI )
 import Distribution.Solver.Modular.Tree
-         ( FailReason(..), POption(..), ConflictingDep(..) )
+    ( FailReason(..), POption(..), ConflictingDep(..) )
 import Distribution.Solver.Modular.Version
+    ( VR, Ver, showVer, showVR, (.||.) )
+
 import Distribution.Solver.Types.ConstraintSource
+    ( ConstraintSource (..), showConstraintSource )
 import Distribution.Solver.Types.PackagePath
+    ( QPN, Qualified(Q), showQPN )
 import Distribution.Solver.Types.Progress
-import Distribution.Solver.Types.ProjectConfigPath (docProjectConfigPathFailReason)
+    ( Progress(..),
+      SummarizedMessage(..),
+      EntryAtLevel(..),
+      Entry(..),
+      SolverState(..) )
+import Distribution.Solver.Types.ProjectConfigPath
+    ( docProjectConfigPathFailReason)
 import Distribution.Types.LibraryName
+    ( LibraryName(LSubLibName, LMainLibName) )
 import Distribution.Types.UnqualComponentName
-import Text.PrettyPrint (nest, render)
-
-data Message =
-    Enter           -- ^ increase indentation level
-  | Leave           -- ^ decrease indentation level
-  | TryP QPN POption
-  | TryF QFN Bool
-  | TryS QSN Bool
-  | Next (Goal QPN)
-  | Skip (Set CS.Conflict)
-  | Success
-  | Failure ConflictSet FailReason
-
--- | Transforms the structured message type to actual messages (strings).
+    ( unUnqualComponentName )
+
+import Text.PrettyPrint ( nest, render )
+
+renderSummarizedMessage :: SummarizedMessage -> String
+renderSummarizedMessage (SummarizedMsg i) = displayMessageAtLevel i
+renderSummarizedMessage (ErrorMsg s) = s
+
+displayMessageAtLevel :: EntryAtLevel -> String
+displayMessageAtLevel (AtLevel l msg) =
+  let s = show l
+  in  "[" ++ replicate (3 - length s) '_' ++ s ++ "] " ++ displayMessage msg
+
+displayMessage :: Entry -> String
+displayMessage (EntryPackageGoal qpn gr) = "next goal: " ++ showQPN qpn ++ showGR gr
+displayMessage (EntryRejectF qfn b c fr) = "rejecting: " ++ showQFNBool qfn b ++ showFR c fr
+displayMessage (EntryRejectS qsn b c fr) = "rejecting: " ++ showQSNBool qsn b ++ showFR c fr
+displayMessage (EntrySkipping cs) = "skipping: " ++ showConflicts cs
+displayMessage (EntryTryingF qfn b) = "trying: " ++ showQFNBool qfn b
+displayMessage (EntryTryingP qpn i) = "trying: " ++ showOptions qpn [i]
+displayMessage (EntryTryingNewP qpn i gr) = "trying: " ++ showOptions qpn [i] ++ showGR gr
```

Comment:
I think that both of these lines should use `showOption` instead of `showOptions`, since they both show single versions.

Reply:
Done!

Reply:
Fixed.